### ISSUE TITLE
Drill Review, Completion, and Assessment for Players

### ISSUE THREAD

**Issue by artkolpakov** [Golf-Drill-Challenge-App/Golf-App#6]
## User Story
As a golf player on the OSU golf team, I want to practice my skills with phone-accessible golf drills assigned to me by my coach, so that I can improve my skills outside of regular team practice.

### Persona
Robert is a 19-year-old golf player at the Oregon State University Golf team who has a smartphone and wants to enhance his golf skills outside of regular team practice sessions. However, Robert does not have a specific training program and tools to track and store associated training data. Additionally, Robert desires the ability to review past drill performances to track his progress.
### Why do we want it
Under the current traditional OSU Men's Golf Team practice methods, players cannot use their phones to follow drills assigned by their coach, restricting their ability to practice asynchronously outside of regular team sessions. Moreover, there is no dedicated platform for the OSU Golf Team to track, store, and share golf drill data, which limits team members from identifying weaknesses and improving their golf skills effectively. Therefore, as part of addressing this issue, we aim to implement a set of features that will provide players with the necessary tools to practice and enhance their golf skills asynchronously, following personalized drills provided by the coaching staff.
### Who is this for
The persona description provided above aims to represent the typical player that this solution seeks to assist, addressing their needs. The tasks and subtasks to be completed as part of this issue all aim to support OSU Golf team players in advancing their golf skills and succeeding in their practice by digitizing their training process.
### Data to support it
The coaching staff of the OSU Golf team has provided our team with the list of drills they intend to incorporate into our app. Consequently, the data metrics for tracking drill performance have also been provided and can be accessed in the following [spreadsheet](https://docs.google.com/spreadsheets/d/1ZufXjRCTzbIqJln_fJ5NDvNo0ahHIM-bkv9ZB8a6Drw/edit#gid=689568128). Therefore, the development team needs to align the application’s data management logistics with the provided drill details and metrics to help the golf team measure drill completion performance.

### Functional

- Displaying drill details information – users with player status should be able to access and view the assigned drill information, including the drill description, completion due date, attached media items, and tracked metrics as part of the drill completion data.
- Inputting drill data upon completion – users with player status should be able to enter and save relevant drill data after completing a specific drill. The input fields will include all the metrics that a particular drill aims to measure, such as distance/proximity to holes in yards or the number of approaches/putts to complete the drill.
- Displaying drill completion data – users with player status should be able to access specific drill data saved after completion. Specifically, players should be able to review the list of completed drills along with the performance data associated with them.
### Non-Functional
- All actions related to inputting/saving/submitting drill result data must be immediately synchronized with the database storage.
User interactions with accessing drill information or inputting drill completion data must have a response time of less than 2 seconds.
- User interactions with accessing drill information or entering drill completion data must exhibit the same efficiency on any mobile device (both iOS and Android).
- The entire user interface for working with the drill data must be scalable and adaptable to any screen size, ensuring accessibility and user-friendliness, with consideration for users with disabilities.
- All drill data, along with interactions with it, must be available for access 24 hours a day, 7 days a week, except during scheduled app maintenance periods.
## Design resources
When implementing the screens related to drills, the styling presented in the following [Figma file ](https://www.figma.com/file/8fP7vgmQ9kNhYeyFIS5guP/OSU-Golf-App-Prototype?type=design&node-id=0%3A1&mode=design&t=Tyms3wWmZjUXt2de-1)must be accurately reflected. Additionally, [React Native’s StyleSheet](https://reactnative.dev/docs/stylesheet) must be used for styling each screen.
### Blocked By
The [Drill Creation for Coaches #4](https://github.com/Golf-Drill-Challenge-App/Golf-App/issues/4) user story is the primary dependency of the current user story. The players would be unable to access any assigned drill data if it doesn't exist. The initial goal is to implement the functionality for the coaches to create drills, including detailed descriptions and associated media items, and subsequently assign them to players. Until this functionality is implemented, the ability for players to access drill data, complete drills, and input performance data upon drill completion is not possible. Furthermore, the implementation of the user sign-up and log-in system is the overarching dependency for all other user stories. The app must store user information to enable the implementation of any other functionalities associated with user interaction within our application.
### Blocking
The [Leaderboard View for Players #5](https://github.com/Golf-Drill-Challenge-App/Golf-App/issues/5) user story is directly dependent on our completion of this issue. This is because leaderboards are generated based on the drill result data entered by the players. Until the functionality for players to review and complete drills, as well as enter their performance results, is implemented, the development of leaderboard-related functionality is completely blocked. However, considering that some of the drill result data might be mocked, one can initiate work on the UI implementation of the leaderboard screen from a proof-of-concept perspective.
#### Implement the “Review Assigned Drills” screen

- [ ] UI Implementation – Implement the interface to showcase the user a list of drills assigned to them by their coach. Each list item will display the drill name and the first two sentences of the drill’s description. Enable drill selection, redirecting the player to the drill details screen upon clicking.
- [ ] UI Styling – Ensure a flawless design for user-friendliness and accessibility. Implement state visualization handlers, including pop-up messages for successful data retrieval, handling network errors, and displaying a loading spinner during the list of drills retrieval.
- [ ] Backend Implementation – Implement the GET API call logic to retrieve the list of drills and associated data for a particular user. Integrate the API call within the frontend logic, mapping the retrieved data to the implemented UI.

#### Implement the “Drill Details” screen
- [ ] UI Implementation part1 – Implement the interface to showcase the user all the details associated with the drill assigned by their coach, including the drill description, metrics for completion assessment, completion deadline, and media items attached by the coach.
- [ ] UI Implementation part2 – Implement the UI logic enabling players to start drill completion, presenting a pop-up window for inputting practice data as they progress through the drill (e.g., data associated with each of the holes). After inputting data, allow players to submit results, closing the pop-up and redirecting to the Drill History screen.
- [ ] UI Styling – Ensure flawless design for user-friendliness and accessibility. Implement state visualization handlers, including pop-up messages for successful drill input, saving data upon submission, and handling input errors and network errors.
- [ ] Backend Implementation – Implement the POST request to save the entered drill completion data. Integrate the API call within the frontend to reflect the corresponding data saving upon drill submission/completion.

#### Implement the “Drill Performance” screen
- [ ] UI Implementation – Implement the interface to enable players to access their personal previous drill performance results obtained from completing drills. The screen should resemble the assigned drills screen, presenting the user with a list of drills. However, this time, players will see their overall performance history within each list item, alongside drill name and description.
- [ ] UI Styling – Ensure flawless design for user-friendliness and accessibility. Implement state visualization handlers, including pop-up messages for successful data retrieval, handling network errors, and displaying a loading spinner during the retrieval of drill performance history. Implement logic for navigating back to the previous screen, as no screen interactions would redirect the user anywhere besides going back; this screen is a dead end.
- [ ] Backend Implementation – Implement the GET API call logic to retrieve the list of drills and associated completion data results for that drill for the particular user. Integrate the API call within the frontend logic, mapping the retrieved data to the implemented UI.

## Estimation
The subtasks for the "Review Assigned Drills" screen implementation are estimated to take 3 story points each (3*3=9). The "Drill Details" screen implementation subtasks sum up to 11 story points (where UI Implementation part 1 is worth 3 story points and part 2 is worth 2 story points). Finally, the "Drill Performance" screen subtasks take 3 story points each as well, summing up to 9. Upon the addition of the story points estimation for the implementation of each screen, we get a total of 29 story points for this user story. Considering that each of the 3-story point tasks is supposed to take no more than a week of work, this evaluates to approximately ~10 weeks of work needed to address this user story, assuming that tasks are being worked on consecutively by a single person. Upon collaborative completion of the issue, 10 weeks could be reduced to 4-5 weeks to complete every single screen.

## Open Questions
- Should players be given a time requirement within which to complete the drill? In other words, once a player starts drill completion on the drill details screen, should they see a timer prompting them to finish the drill and submit all performance data within a specified duration (similar to Canvas quizzes)? 
- Do we intend to support due dates for drill completion? If a player fails to finish the drill by the due date, should they be automatically unassigned from the drill, or do we allow late submissions and notify the coach about them?
- Should players have the option to recomplete a specific drill? For instance, if a player is dissatisfied with their performance, can they restart the drill completion and reenter their performance data? Do we prefer to overwrite the drill performance history in such cases, or should we track the number of attempts for drill completion with corresponding data results?
- Should players be permitted to unassign themselves from drill completion within the drill details screen? This would involve notifying the coach and providing a reason for not completing the drill.
- Should players have the ability to delete individual drill data or their entire drill history when reviewing their performance on the drill performance screen?

## Acceptance Criteria
Both the UI and backend logic for the "Review Assigned Drills" screen, "Drill Details" screen, and "Drill Performance" screen have been implemented. The UI workflow enables users to review the list of drills assigned to them and access each drill's details individually from that list. Additionally, users can complete drills assigned to them by starting them and entering performance results. After completion, users can access their performance records history. The screens' design is user-friendly and accessible, and interactions with the database are efficient and quick.

Using the behavior-driven development _"Given/When/Then"_ format:

- **Given** a user is on the "Review Assigned Drills" screen, **When** a user taps on the drill they want to start, **Then** the user gets redirected to the drill details screen.
- **Given** a user is on the “Drill Details”, **When** a user taps on the “start” button, **Then** the user is prompted to a pop-up that has all the metrics for tracking the drill completion.
- **Given** a user is on the “Drill Details” and they have started completing the drill, **When** a user enters all the data and presses the “submit” button, **Then** the success message shows up, and the user gets redirected to the “Drill Performance” screen showing their saved performance results.
- **Given** a user is on the “Drill Performance” screen and they reviewed all the performance data they wanted to see, **When** the user chooses to navigate back, **Then** they get navigated to the previous screen they were on.

## Definition of Done (DoD)
- The produced implementation aligns with the acceptance criteria outlined in the issue description, ensuring a precise match with all Given/When/Then scenarios.
- The code attempting to complete the task is fully documented, including comment headers for every function and comments explaining the logic of every 10 lines of code.
- The code attempting to complete the task achieves a minimum of 90% test coverage and does not cause any existing unit tests to fail.
- The code attempting to complete the task receives two approvals on its pull request, and its remote feature branch is squash-merged into the main branch and subsequently deleted.
- Both functional and non-functional requirements are met, where functional testing is conducted by all team members, and no regressions are observed.
- The application’s documentation is updated to address the added screens and the functionality they provide to the user.
- All of the newly added screens undergo a demonstration for the OSU Golf team’s players and coaches, successfully passing user acceptance testing. During this process, players engage in hands-on exploration, reviewing and interacting directly with the newly introduced screens, confirming that all functionalities align with their expectations.
